shared/src/main/scala/Interpreter.scala

```diff
@@ -126,24 +129,84 @@ object Interpreter:
       args: Seq[VAny] | Null = null,
       popArgs: Boolean = true
   )(using ctx: Context): VAny =
-    val VFun(impl, arity, params, origCtx) = fn
+    val VFun(impl, arity, params, origCtx, origAST) = fn
+    val useStack = arity == -1
+    val vars: mut.Map[String, VAny] = mut.Map()
     val inputs =
-      if args != null then args
-      else if popArgs then ctx.pop(arity)
-      else ctx.peek(arity)
+      if args != null && params.isEmpty then args
+      else if arity == -1 then List.empty // operates on entire stack
+      else if params.isEmpty then // no params, so just pop the args
+        if popArgs then ctx.pop(arity) else ctx.peek(arity)
+      else
+        var argIndex: Int = 0
+        val origLength = ctx.length
+        val popFunction = (n: Option[Int]) =>
```

Comment:
```suggestion
        def popFunction(n: Option[Int]): VAny | Seq[VAny] =
```
Since the result type is a union type, I think you need to explicitly give the result type or it'll infer `Any`.